### PR TITLE
Support * Wildcard for quarkus.http.cors.origins

### LINKED ISSUE #5422: quarkus.http.cors.origins=* should be supported
**Describe the bug**
When I read the documentation about configuring cors for the first time, I tought that leaving "quarkus.http.cors.origins" as empty would be the same as "Access-Control-Allow-Origin: *".
The * wildcard allows any origin to access ressources.

Furthermore, a value of "*" is currently not supported. See CORSFilter.java#L93.
`boolean allowsOrigin = corsConfig.origins.isEmpty() || corsConfig.origins.contains(origin);`

We should try to improve the documentation to reflect what is actually going on in the filter.

**Expected behavior**

> For requests without credentials, the literal value "*" can be specified, as a wildcard; the value tells browsers to allow requesting code from any origin to access the resource.

If I use "*" as wildcard, I expect quarkus to allow any origin to access my resources.
If I remove "quarkus.http.cors.origins" in my configuration, I expect quarkus to use "*" as default.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin

**Actual behavior**
Requests get rejected, altough quarkus.http.cors.origins is set to "*".

**Configuration**
```properties
uarkus.http.cors=true
quarkus.http.cors.origins=*

```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
SYS_NT-10.0 DESKTOP-0TKUM9I 2.11.2(0.329/5/3) 2018-11-10 14:38 x86_64 Msys
 - Output of `java -version`:
java version "1.8.0_211"
Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.211-b12, mixed mode)
 - Quarkus version or git rev: 1.0.0.CR1

### PR DESCRIPTION
**Pull request by Postremus**
The star wildcard for the origins means, that any origin can access the service.
Previosly, this behavior could only be achieved by leaving the config property out completly.
Now you can either not set the origins config property, or use "*".

Adjust the documentation for * origins.
Fix naming of one of the tests.

fixes #5422

### CODE REVIEW

File: extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java
Comment (machi1990) [inline extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java]: Good catch.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: @Postremus Can you please  check it is a single value list too and use `get(0).equals` ? We probably should not allow some random lists which might have non wildcard and wildcard values :-)
There also should be a unit test, please copy and paste one of the tests there to test a `*`

File: extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java
Comment (sberyozkin) [inline extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java]: @Postremus thanks for spotting the typo, it is a minor issue but lets keep the original naming pattern, `CorsSecurityTestCase`

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: I'm not sure it should be a default value - that would mean all other properties should also be defaulted to `*`

File: extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java
Comment (Postremus) [reply extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java]: @sberyozkin The other test classes in this package are also named CORSXXXXTestCase, which is why I went for this name.

File: docs/src/main/asciidoc/http-reference.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/http-reference.adoc]: See my note about the default value. Perhaps it should read: `The filter allows any origin if this is not set or set to '*'' ?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (machi1990) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: We should probably revert this and allowsAllOrigin if the list contains only `*` or `isEmpty()`

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: @machi1990 yes, I agree

Comment (Postremus) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: We may need to allow `*` for more of the cors properties anyway in the future.
`Access-Control-Expose-Headers`, `Access-Control-Allow-Methods` and `Access-Control-Allow-Headers` allow * as wildcard value.

Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: @Postremus sure, it makes sense for the consistency purposes. But lets keep the current pattern, it is either absent or set to `*`

**Comment by sberyozkin** [comment]
@Postremus Hi, can you please do `mvn process-sources` in `quarkus-vertx-http` ?

As mentioned above, for the sake of consistency, all other CORS properties which were allowing for a wildcard when omitted should also be updated.

I have to admit I'm a bit not sure now what is the best approach, the one your originally started with (property default to `*`) or the one I proposed yesterday (keep the original pattern). @machi1990, should we ask @Postremus to continue with his original idea ? It seemed to me late yesterday that it was changing things, but now, I'm thinking, if a given property is omitted => it is an indirect `*` or it can be defaulted to `*` from the start, in which a given property is never omitted OOB, yet it does not change anything, it is still `*` by default :-).
CC @gsmet

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: Hmmm, isn't it a breaking change? If it's empty, it used to pass, now it doesn't. Is it really what we want? Or did I miss something.

Blocking merge until I got an answer to that.

**Review by gsmet — CHANGES_REQUESTED** [review]

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: Good catch, `corsConfig.origins.isEmpty()` should still be there.

**Comment by sberyozkin** [comment]
@Postremus 
I'm sorry for my rather confusing review, I did it in a great rush. So, here is a proposal: please revert to your original idea by defaulting to `*` as it makes things simpler without breaking the compatibility, only enforce that a list containing a `*` is a single element list. And make sure that all other similar properties are supported consistently. Give it a go please when you get a chance

**Comment by gsmet** [comment]
I would have liked to have this one in 1.0.0.Final. Can someone finish it before Friday end of day?

**Comment by Postremus** [comment]
@sberyozkin sorry for not responding earlier.
@gsmet I will give it a go.

**Comment by machi1990** [comment]
> @sberyozkin sorry for not responding earlier.
> @gsmet I will give it a go.

Thanks @Postremus

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: Why this change?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: I don't understand this change?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: I asked for it to support the empty list and I don't think you have addressed it.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a few comments/questions, I don't think we will converge on that one today.

Let's target the next release.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (Postremus) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: We can't use an Enum here, when we want to support * wildcard for methods as well.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (Postremus) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: This change is done so that we also support * for methods as well. See new diff.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (Postremus) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: done

**Comment by gsmet** [comment]
Oh and I just stumbled upon the fact that this will conflict with https://github.com/quarkusio/quarkus/pull/5387/commits . We want to get this one in the tree ASAP as it's a very large one and it's a pain to rebase.

We will have to revisit that one after the merge.

**Comment by machi1990** [comment]
@Postremus I noticed a formatting issue in the CI, could you run `mvn formatter:format` in the `vertx-http` module? and force push your changes. Thanks.

**Comment by Postremus** [comment]
@machi1990 done.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: @Postremus sorry I'm getting back to this one a bit late. I rebased and added a commit to improve the documentation formatting.

And while doing so, I saw this change. It looks like a significant change in behavior or am I wrong? Did it really expose all headers by default before?

**Review by gsmet — CHANGES_REQUESTED** [review]
I asked an additional question.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java
Comment (gsmet) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: @Postremus ping?

Comment (Postremus) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSConfig.java]: @gsmet pong :)
You are right. That is quite the change from previous behavior. I removed this change.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: You could remove the first part of the `if`.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: You could remove the first part of the `if`.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: I think you mistakenly reverted the config changes necessary for this to work?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: I don't understand why this is necessary?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (gsmet) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java]: You already checked the value is present. No need for a `orElse` (I know it was preexisting but let's get rid of it).

**Review by gsmet — CHANGES_REQUESTED** [review]
I added some comments inline.

I think you reverted some config changes that are needed. Unless I miss something.

**Comment by sberyozkin** [comment]
Closing this one after #12142 has been merged, @Postremus thanks for starting this effort earlier

**Pull request closed without merging**
